File: js/src/errors/query-run-errors.ts
import { BaseError } from "./base-error";
import { ERROR_TYPES } from "./error-types";

export class QueryRunRateLimitError extends BaseError {
  constructor() {
    const errorType = ERROR_TYPES.query_run_rate_limit_error;
    super(`${errorType}: you have exceeded the rate limit for creating/running new queries.`);
    this.errorType = errorType;
  }
}

export class QueryRunTimeoutError extends BaseError {
  constructor(timeoutMinutes: number) {
    const errorType = ERROR_TYPES.query_run_timeout_error;
    super(`${errorType}: your query has timed out after ${timeoutMinutes} minutes.`);
    this.errorType = errorType;
  }
}

export class QueryRunExecutionError extends BaseError {
  constructor({
    name,
    message,
    data,
  }: {
    name?: string | undefined | null;
    message?: string | undefined | null;
    data?: Record<any, any> | undefined | null;
  }) {
    const errorType = ERROR_TYPES.query_run_execution_error;
    if (!name && !message && !data) {
      super(`${errorType}: an error has occured while executing your query.`);
    } else {
      super(
        `${errorType}: an error has occured while executing your query: name=${name} - message=${message} - data=${data}`
      );
    }
    this.errorType = errorType;
    if (name) {
      this.name = name;
    }
    if (message) {
      this.message = message;
    }
    if (data) {
      this.data = data;
    }
  }
}

File: python/src/errors/query_run_errors.py
from typing import Optional, Union

from .base_error import BaseError


class QueryRunRateLimitError(BaseError):
    """
    Base class for all QueryRunRateLimitError errors.
    """

    def __init__(self):
        self.message = "QUERY_RUN_RATE_LIMIT_ERROR: you have exceeded the rate limit for creating/running new queries"
        super().__init__(self.message)


class QueryRunTimeoutError(BaseError):
    """
    Base class for all QueryRunTimeoutError errors.
    """

    def __init__(self, timeoutMinutes: Union[int, float, None] = None):
        if timeoutMinutes is None:
            self.message = f"QUERY_RUN_TIMEOUT_ERROR: your query has timed out."
        else:
            self.message = f"QUERY_RUN_TIMEOUT_ERROR: your query has timed out after {timeoutMinutes} minutes."
        super().__init__(self.message)


class QueryRunExecutionError(BaseError):
    """
    Base class for all QueryRunExecutionError errors.
    """

    def __init__(
        self,
        error_name: Optional[str] = None,
        error_message: Optional[str] = None,
        error_data: Optional[str] = None,
    ):
        self.message = f"QUERY_RUN_EXECUTION_ERROR: an error has occured while executing your query. errorName={error_name}, errorMessage={error_message}, errorData={error_data}"
        super().__init__(self.message)


class QueryRunCancelledError(BaseError):
    """
    Base class for all QueryRunCancelledError errors.
    """

    def __init__(
        self,
        error_name: Optional[str] = None,
        error_message: Optional[str] = None,
        error_data: Optional[str] = None,
    ):
        self.message = f"QUERY_RUN_CANCELLED_ERROR: your query has been cancelled. errorName={error_name}, errorMessage={error_message}, errorData={error_data}"
        super().__init__(self.message)


class QueryRunInvalidStateToCancel(BaseError):
    """
    Base class for all QueryRunInvalidStateToCancel errors.
    """

    def __init__(
        self,
        msg: Optional[str] = None,
    ):
        self.message = msg
        super().__init__(self.message)
